### DISCUSSION TITLE
Airflow pods stuck in running

### DISCUSSION THREAD

**Discussion by AdrianCio** [apache/airflow#29574] [Q&A]
Hi!

Sometimes, seemingly random, it happens that pods created by Airflow via `GKEStartPodOperator` are getting stuck in `Running`. These pods will simply stay there indefinitely. We encountered different variations of this:

- base container is stuck in `running`
- base container is `completed` but sidecar is `running`

The affected pods show `containers with unready status: [base]` message in the GCP UI. Maybe worth mentioning that even successful ones will show up for a while as `pod not ready` before eventually closing successfully after 1-2 minutes.

We are trying to pinpoint the issue for a couple of days now and we're running out of options. 

We use `composer-1.20.1-airflow-2.3.4` deployed in GCP. The version we use in python is `apache-airflow==2.3.4`. 
The node pool on which Airflow is running the jobs has autoscaling enabled. 

We could find some correlation between the moments when autoscaler is increasing/decreasing and the pod start time for those that get stuck in `Running`, but we also ran an experiment where we forced max 2 tasks to be ran at a time which kept the node pool at a steady level so autoscaler did not kick in, yet we still had two pods hang (both started at the same time), but overall it seemed much less error-prone.

This is an example of how it looks like:

Note that we don't define any limit to the resources. Would that be a possible cause for the issues?
```
GKEStartPodOperator(
    task_id=id,
    do_xcom_push=True,
    # k8s configuration
    name=f"some-na,e",  # used to generate Pod ID
    project_id=(config["gcp-project-id"]),
    cluster_name=(config["k8s-cluster"]),
    location=(config["gcp-location"]),
    is_delete_operator_pod=True,
    # task input
    env_vars={
        MY_VAR=""
    },
    arguments=[
        "python",
        "main.py",
        # enable debug, xcom (+ passthroughs)
        "-dx"
    ],
    # Pod specification
    namespace="airflow",
    labels={"app": "airflow-dag"},
    image=config["workloads"]["docker-image"],
    image_pull_policy="IfNotPresent",
    volumes=[get_secret_volume("airflow")],
    volume_mounts=[get_volume_mount("airflow")],
    container_resources=k8s_models.V1ResourceRequirements(
        requests={
            "memory": config["workloads"]["predictor"]["request"][
                "memory"
            ],
            "cpu": config["workloads"]["predictor"]["request"]["cpu"],
        },
    ),
    # Fail automatically after 1 hour of runtime
    execution_timeout=datetime.timedelta(
        seconds=3600
    ),
    startup_timeout_seconds=(
        config["workloads"]["predictor"]["startup-timeout-seconds"]
    ),
    affinity=node_affinity,
)
```

For a particular DAG it happens much more often but we encountered one-off situations in other DAGs also.

**Comment by hussein-awala** [comment]
Airflow `GKEStartPodOperator` extends `KubernetesPodOperator` which is used by a lot of companies without any problem, the difference between the two operators is the cluster configuration which is not related to your issue. So if there is a problem in the GKE operator, we should be able to reproduce with `KubernetesPodOperator`.

Can you check what you get when you call:
```bash
kubectl describe pod <pod name>
```

**Reply by AdrianCio** [reply]
Example 1: pot stuck in Running

The base container:
```
State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Thu, 16 Feb 2023 07:21:52 +0100
      Finished:     Thu, 16 Feb 2023 07:30:58 +0100
Ready:          False
Restart Count:  0
```

Airflow-xcom-sidecar
```
Command:
      sh
      -c
      trap "exit 0" INT; while true; do sleep 1; done;
State:          Running
    Started:      Thu, 16 Feb 2023 07:21:52 +0100
Ready:          True
Restart Count:  0
```

The pod itself:
```
Conditions:
Type              Status
Initialized       True 
Ready             False 
ContainersReady   False 
PodScheduled      True 
```

Example 2 pod stuck in Completed:

Base container is the same as above

airflow-xcom-sidecar:
```
State:          Terminated
    Reason:       Completed
    Exit Code:    0
    Started:      Thu, 16 Feb 2023 07:07:32 +0100
    Finished:     Thu, 16 Feb 2023 07:07:34 +0100
Ready:          False
Restart Count:  0
```

The pod itself:
```
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
```

There are no events in any of them.

What stood out is that the sidecar container starts a couple of seconds after the base but I'm not sure whether that matters.

**Reply by hussein-awala** [reply]
can you provide the `pod_manager.py` and `kubernetes_pod.py`? (the lines in the log logged form these two files)
